contextualized/regression/datasets.py

```diff
@@ -1,121 +1,120 @@
 """
-Data generators used for Contextualized regression training.
+Datasets used for Contextualized regression training.
 """
 from abc import abstractmethod
 import torch
-from torch.utils.data import IterableDataset
 
 
-class Dataset:
-    """Superclass for datastreams (iterators) used to train contextualized.regression models"""
+class RegressionDatasetBase:
+    """
+    Superclass for map-based datasets used to train contextualized.regression models
+    """
+
+    def __init__(self, C, X, Y, idx=[], dtype=torch.float):
 
-    def __init__(self, C, X, Y, dtype=torch.float):
         self.C = torch.tensor(C, dtype=dtype)
         self.X = torch.tensor(X, dtype=dtype)
         self.Y = torch.tensor(Y, dtype=dtype)
-        self.n_i = 0
-        self.x_i = 0
-        self.y_i = 0
-        self.n = C.shape[0]
         self.c_dim = C.shape[-1]
         self.x_dim = X.shape[-1]
         self.y_dim = Y.shape[-1]
+        self.n = self.C.shape[0]
         self.dtype = dtype
 
-    def __iter__(self):
+        if idx == []:  # use full dataset
+            self.idx = range(self.n)
+        else:
+            self.idx = idx
+
+        self._generate_samples()
+
+    def _generate_samples(self):
         self.n_i = 0
-        self.x_i = 0
         self.y_i = 0
-        return self
+        self.x_i = 0
+        self.all_samples = []
+
+        for i in range(self._total_len()):
```

Comment:
This seems like the only place `_total_len` is used. Should probably be replaced with `len(self)` if we can guarantee that both should always return the same value.